File: api/src/main/kotlin/edu/wgu/osmt/api/model/ApiSkillUpdate.kt
package edu.wgu.osmt.api.model

import com.fasterxml.jackson.annotation.JsonProperty
import edu.wgu.osmt.db.PublishStatus

data class ApiSkillUpdate(
    @JsonProperty("skillName")
    override val skillName: String? = null,
    
    @JsonProperty("skillStatement")
    override val skillStatement: String? = null,
    
    @JsonProperty("status")
    override val publishStatus: PublishStatus? = null,
    
    @JsonProperty("collections")
    override val collections: ApiStringListUpdate? = null,
    
    @JsonProperty("authors")
    val authors: ApiStringListUpdate? = null,
    
    @JsonProperty("categories")
    val categories: ApiStringListUpdate? = null,
    
    @JsonProperty("keywords")
    override val keywords: ApiStringListUpdate? = null,
    
    @JsonProperty("certifications")
    override val certifications: ApiReferenceListUpdate? = null,
    
    @JsonProperty("standards")
    override val standards: ApiAlignmentListUpdate? = null,
    
    @JsonProperty("alignments")
    override val alignments: ApiAlignmentListUpdate? = null,
    
    @JsonProperty("employers")
    override val employers: ApiReferenceListUpdate? = null,
    
    @JsonProperty("occupations")
    override val occupations: ApiStringListUpdate? = null
) : SkillUpdate {
    
    fun validate(rowNumber: Number? = null): List<ApiFieldError>? {
        val errors = mutableListOf<ApiFieldError>()
        return if (errors.size > 0) errors else null
    }
    
    fun validateForCreation(rowNumber: Number? = null): List<ApiFieldError>? {
        val errors = mutableListOf<ApiFieldError>()
        
        if (skillName.isNullOrBlank()) {
            errors.add(ApiFieldError(field = "skillName", message = "Name is required", rowNumber = rowNumber))
        }
        if (skillStatement.isNullOrBlank()) {
            errors.add(
                ApiFieldError(
                    field = "skillStatement",
                    message = "Statement is required",
                    rowNumber = rowNumber
                )
            )
        }
        
        validate()?.let { errors.addAll(it) }
        
        return if (errors.size > 0) errors else null
    }
}

interface SkillUpdate {
    val skillName: String?
    val skillStatement: String?
    val publishStatus: PublishStatus?
    val collections: ApiStringListUpdate?
    val keywords: ApiStringListUpdate?
    val certifications: ApiReferenceListUpdate?
    val standards: ApiAlignmentListUpdate?
    val alignments: ApiAlignmentListUpdate?
    val employers: ApiReferenceListUpdate?
    val occupations: ApiStringListUpdate?
}

File: ui/src/app/richskill/ApiSkillUpdate.ts
import {ApiAlignment, ApiNamedReference, IAlignment, INamedReference} from "./ApiSkill"
import {PublishStatus} from "../PublishStatus"

export interface IStringListUpdate {
  add?: string[]
  remove?: string[]
}
export class ApiStringListUpdate implements IStringListUpdate {
  add?: string[]
  remove?: string[]

  constructor(add?: string[], remove?: string[]) {
    this.add = add
    this.remove = remove
  }
}

export interface IReferenceListUpdate {
  add?: INamedReference[]
  remove?: INamedReference[]
}
export class ApiReferenceListUpdate {
  add?: ApiNamedReference[]
  remove?: ApiNamedReference[]

  constructor(add?: ApiNamedReference[], remove?: ApiNamedReference[]) {
    this.add = add
    this.remove = remove
  }
}

export interface IAlignmentListUpdate {
  add?: ApiAlignment[]
  remove?: ApiAlignment[]
}
export class ApiAlignmentListUpdate {
  add?: ApiAlignment[]
  remove?: ApiAlignment[]

  constructor(add?: ApiAlignment[], remove?: ApiAlignment[]) {
    this.add = add
    this.remove = remove
  }
}



export interface IRichSkillUpdate {
  skillName?: string
  skillStatement?: string
  status?: PublishStatus
  category?: string
  keywords?: IStringListUpdate
  collections?: ApiStringListUpdate
  alignments?: IAlignmentListUpdate
  standards?: IAlignmentListUpdate
  certifications?: ApiReferenceListUpdate
  occupations?: ApiStringListUpdate
  employers?: ApiReferenceListUpdate
  authors?: IStringListUpdate
}

export class ApiSkillUpdate implements IRichSkillUpdate {
  skillName?: string
  skillStatement?: string
  status?: PublishStatus
  category?: string
  keywords?: ApiStringListUpdate
  collections?: ApiStringListUpdate
  alignments?: ApiAlignmentListUpdate
  standards?: ApiAlignmentListUpdate
  certifications?: ApiReferenceListUpdate
  occupations?: ApiStringListUpdate
  employers?: ApiReferenceListUpdate
  authors?: ApiStringListUpdate

  constructor({skillName, skillStatement, status, category, keywords, collections, alignments, certifications, standards, occupations, employers, authors}: IRichSkillUpdate) {
    this.skillName = skillName
    this.skillStatement = skillStatement
    this.status = status
    this.category = category
    this.keywords = keywords
    this.collections = collections
    this.alignments = alignments
    this.certifications = certifications
    this.standards = standards
    this.occupations = occupations
    this.employers = employers
    this.authors = authors
  }
}
